### DISCUSSION TITLE
Clarification on Integrating New Features into Robotics Infrastructure Repository

### DISCUSSION THREAD

**Discussion by izzaaaatanishq** [JdeRobot/RoboticsAcademy#2437] [Q&A]
Hello everyone,

I hope this message finds you well. I've encountered a query regarding the project's directory structure and the incorporation of new features.

I've observed that the robots are managed within a separate Git repository distinct from the Robotics Academy's repository. In light of this, would it be advisable for us to integrate any new features directly into the Robotics Infrastructure repository?

Furthermore, I'm curious about how these changes will be reflected in the exercises. Given that modifications will be made locally within the repository, what would be the best practice to ensure these changes are effectively synchronized with the exercises?

I appreciate any insights or guidance you can provide on this matter. Thank you for your time and assistance.

Best regards,
Tanishq

**Comment by jmplaza** [comment]
All the robot models, scenarios, launch files etc.. are stored in [RoboticsInfrastructure repo](https://github.com/JdeRobot/RoboticsInfrastructure). This way they can be used from other tools such as [VisualCircuit](https://github.com/JdeRobot/VisualCircuit) or [BT-Studio](https://github.com/JdeRobot/bt-studio).

Yes, if you improve a robot model or scenario for one RoboticsAcademy (RA) exercise the changes should be located at RoboticsInfrastructure repo. Also new robot models and new Gazebo worlds are also welcome. From RoboticsAcademy they will be referenced as a **robotics universe**. Each RA exercise has a robotics universe at least. It may even have several ones, for instance in FollowLine exercise several different circuits. They are referenced in the RA exercise database.

Hope it helps,

**Reply by izzaaaatanishq** [reply]
Can you mention how they are referencing those models from different repository?

**Reply by jmplaza** [reply]
In the "configuration" field at the exercise table of RoboticsAcademy. Anyway we are refactoring this... It will change a little bit in the upcoming RoboticsAcademy 5.2. Check the [developer documentation](https://github.com/JdeRobot/RoboticsAcademy/blob/master/docs/InstructionsForDevelopers.md), although it may be a little bit outdated.

**Reply by izzaaaatanishq** [reply]
Hi @jmplaza,

I've reviewed the documentation thoroughly but couldn't locate any information on the robot infrastructure. I did come across one method, but I'm uncertain if it's advisable.

The method involves forking the robot infrastructure repository to my account, making the necessary modifications, and then pushing the changes to a new branch. During the process of building the new image, the link to the newly forked repository with the new branch would be altered in the build.sh file within Mini Radi.

Is this approach acceptable, or is there a better alternative?

Method on Docs: [Steps to change models from CustomRobots in RoboticsAcademy exercises.](https://github.com/JdeRobot/RoboticsAcademy/blob/master/docs/InstructionsForDevelopers.md#steps-to-change-models-from-customrobots-in-roboticsacademy-exercises)


So shouldn't there be a method to work with models also? like we do with the robot academy repo by connecting it through web sockets.

**Reply by pawanw17** [reply]
Hi @Tanishq30052002 During the building of a miniRADI the Robotics Infra repo is cloned, if you want to test your own changes, you can
1. push to your fork and mention the fork link in the Dockerfile of the image you plan to build (noetic or humble https://github.com/JdeRobot/RoboticsAcademy/blob/3bf19e0d778a11479be6b4c20d788ca6118b9e1a/scripts/mini_RADI/Dockerfile.mini_noetic#L11)
and while running the build.sh place your branch name in the parameters (https://github.com/JdeRobot/RoboticsAcademy/blob/master/docs/generate_a_mini_radi.md#usage)

2. You could also copy your RoboticsInfra repo after an image is already built by using "docker cp" command, but post that you'll have to do a few steps that the Dockerfile does for you manually.


I generally use the 1st method.

**Reply by izzaaaatanishq** [reply]
Hi @pawanw17 ,
But don't you think the 1st method is actually slow?
Since we need to build the image every time we do a simple change.
Can't we just do the same thing with it also as we do with robotics academy repo.

**Reply by pawanw17** [reply]
The method I posted was for a considerably large change/ final testing before making a PR. When I need to do quick testing of my changes, I simply open a RADI in debug mode (--debug ) flag, make changes and 
run the application using 
./entrypoint.sh
if things turn out well then great else I press Ctrl + c and run ./kill.sh (only present in Humble RADI afaik) and inspect what went wrong

**Reply by pawanw17** [reply]
I also often use docker exec to open multiple terminals to a RADI for testing and debugging purposes

`docker exec -ti <container name> bash`

you can find the container name using
`docker ps`

**Reply by izzaaaatanishq** [reply]
so, you mean you do changes in the container itself. Right?

**Reply by pawanw17** [reply]
That's right

**Reply by izzaaaatanishq** [reply]
ok
